rfcs/proposed/task_group_dynamic_dependencies/README.md

```diff
@@ -0,0 +1,366 @@
+newly created subtasks to complete.
+
+<img src="merge_sort.png" width=800>
+
+A key point of this recursive parallel algorithm is the requirement to change 
+the predecessors of the merge tasks. However, the merge tasks are already 
+submitted when their predecessors are modified. As mentioned in the previous 
+section, updating the predecessors of a submitted task can be
+risky due to the potential for a race condition. However, in this case,
+it is safe to add or change predecessors to the merge task. 
+This is because the merge task cannot start execution until all of its current predecessors 
+complete. Those predecessors are the tasks responsible for modifying the merge task dependencies. 
+
+Therefore, we propose a limited extension that allows transferring 
+all the successors of the currently executing task to become the successors 
+of a different created task. This function can only access the successors
+of the currently executing task, and those tasks are prevented from executing 
+by a dependence on the current task itself, so we can ensure that we can safely 
+update the incoming dependencies for those tasks without worrying about
+potential race conditions. 
+
+One possible spelling for this function would be `transfer_successors_to(h)`, 
+Where `h` is a `task_handle` to a created task, and the 
+`transfer_successors_to` function must be called from within a task. Calling
+this function from outside a task or passing anything other than a `task_handle`
+representing a task in the created state is undefined behavior.
+
+### Examples
+
+##### Simple Three-Nodes Task Graph
+
+The example below shows a simple graph with three nodes. 
+`final_task` must wait for  `first_task` and `second_task` 
+to complete.
+
+    tbb::task_group tg;
+
+    tbb::task_handle first_task = tg.defer([&] { /* task body */ });
+    tbb::task_handle second_task = tg.defer([&] { /* task body */ });
+    tbb::task_handle final_task = tg.defer([&] { /* task body */ });
+
+    final_task.add_predecessor(first_task);
+    final_task.add_predecessor(second_task);
+
+    // order of submission is not important
+    tg.run(std::move(final_task));
+    tg.run(std::move(first_task));
+    tg.run(std::move(second_task));
+
+    tg.wait();
+
+The dependency graph for this example is:
+
+<img src="three_task_graph.png" width=400>
+
+#### Predecessors in Unknown States
+
+The example below shows a graph where the dependencies are determined
+dynamically. The state of the predecessors is unknown. The
+`users::find_predecessors` function returns, based on application
+logic, the tasks that must complete before the new work can start.
+
+    void add_another_task(tbb::task_group& tg, int work_id) {
+        tbb::task_handle new_task = tg.defer([=] { do_work(work_id); });
+
+        for (tbb::task_handle& p : users::find_predecessors(work_id)) {
+            new_task.add_predecessor(p);
+        }
+
+        tg.run(std::move(new_task));
+    }
```

Comment:
```suggestion
    tbb::task_handle add_another_task(tbb::task_group& tg, int work_id) {
        tbb::task_handle new_task = tg.defer([=] { do_work(work_id); });

        for (tbb::task_handle& p : users::find_predecessors(work_id)) {
            new_task.add_predecessor(p);
        }
        tg.run(new_task);
        return new_task;
    }
```
The suggested change does not move the created task handle to `run()`, instead returns it to the caller. I think that would be practically necessary for `users::find_predecessors` to work,

Reply:
Am I correct that `new_task` can be useful for the caller in terms of `users::find_predecessors` because it allows to make `new_task` visible for other calls of `add_another_task` and makes it possible to use is as a predecessor for other tasks?

Reply:
Right, that's the idea.